backend/main/shared/dps.go

```diff
@@ -0,0 +1,213 @@
+	}
+)
+
+type DpsAdapter struct {
+	Config  FlowConfig
+	Invoker *invoker.Invoker
+	Context context.Context
+	URL     string
+	Env     string
+}
+
+func NewDpsClient(flowEnv string) *DpsAdapter {
+	// Initialize adapter
+	adapter := DpsAdapter{}
+	adapter.Context = context.Background()
+	adapter.Env = flowEnv
+
+	// parse flow.json
+	path := "./flow.json"
+
+	content, err := ioutil.ReadFile(path)
+
+	if err != nil {
+		log.Fatal().Msgf("Error when opening file: %+v.", err)
+	}
+
+	var config FlowConfig
+	err = json.Unmarshal(content, &config)
+	if err != nil {
+		log.Fatal().Msgf("Error parsing flow.json: %+v.", err)
+	}
+
+	adapter.Config = config
+	adapter.URL = config.DpsNetworks[adapter.Env]
+
+	// For emulator, skip invoker instantiation
+	if adapter.URL == "" {
+		return &adapter
+	}
+
+	// Initialize the DPS API client.
+	codec := zbor.NewCodec()
+
+	conn, err := grpc.DialContext(
+		adapter.Context,
+		adapter.URL,
+		grpc.WithTransportCredentials(insecure.NewCredentials()),
+		grpc.FailOnNonTempDialError(true), // DialContext will fail immediately if passed bad connection string
+		grpc.WithBlock(),                  // connection should be blocking
+	)
+
+	if err != nil {
+		log.Error().Str("dps", adapter.URL).Err(err).Msg("could not dial API host")
+		panic("error initializing DPS gRPC connection")
+	}
+
+	client := dpsApi.NewAPIClient(conn)
+	index := dpsApi.IndexFromAPI(client, codec)
+
+	cacheSize := uint64(100_000_000)
+	_invoker, err := invoker.New(index, invoker.WithCacheSize(cacheSize))
+	adapter.Invoker = _invoker
+
+	if err != nil {
+		log.Error().Str("dps", adapter.URL).Err(err)
+		panic("error initializing DPS Invoker")
+	}
+	return &adapter
+}
+
+// FLOW token has its own function for fetching balance because
+// it has its own script to account for locked/staked tokens.
+func (dps *DpsAdapter) GetFlowBalanceAtBlockheightScript(addr string, blockheight uint64) (cadence.Value, error) {
+	var args []cadence.Value
+
+	// Prepare Args
+	flowAddress := flow.HexToAddress(addr)
+	cadenceAddress := cadence.NewAddress(flowAddress)
+
+	args = append(args, cadenceAddress)
+
+	_script, err := ioutil.ReadFile("./main/cadence/scripts/get_flow_token_balances.cdc")
+
+	if err != nil {
+		log.Error().Err(err).Msgf("Error reading cadence script file.")
+		return nil, err
+	}
+
+	// Replace script args
+	script := dps.Config.InsertCoreContractAddresses(string(_script))
+
+	result, err := dps.Invoker.Script(blockheight, []byte(script), args)
+	if err != nil {
+		log.Error().Err(err).Msg("invoker.Script failed")
+		return nil, err
+	}
+	return result, nil
+}
+
+func (dps *DpsAdapter) GetBalanceAtBlockheight(addr string, blockheight uint64, c *Contract) (*FTBalanceResponse, error) {
+	if dps.bypass() {
+		log.Info().Msgf("overriding dps service for emulator")
+		return &DummyBalance, nil
+	}
+
+	var err error
+	var result cadence.Value
+
+	tokenBalanceResponse := FTBalanceResponse{}
+	tokenBalanceResponse.NewFTBalance()
```

Comment:
maybe this is for a new separate PR.. but `NewFTBalance` will return hardcoded dummy balances if running on the emulator, but I think now with DPS we want to ping DPS on local dev too right?

Reply:
ok got it